### PR TITLE
2PC with Group-By

### PR DESCRIPTION
**Pull request by rithvikp**

### CODE REVIEW

File: hydroflow/examples/two_pc/coordinator.rs [hunk 1]
```diff
@@ -118,38 +134,72 @@ pub(crate) async fn run_coordinator(opts: Opts, subordinates: Vec<String>) {
...
+            .map_scan(0u32, |counter, _| {
+                *counter += 1;
+                *counter
+            })
+            .pull_to_push()
+            .map(Some)
+            .push_to(subordinate_count_1_push),
+    );
+
+    // Take the max of subordinate counts.
```
Comment (jhellerstein) [inline hydroflow/examples/two_pc/coordinator.rs:152]: Can we add a comment to explain the reasoning for MAX(COUNT) here?

Comment (jhellerstein) [reply hydroflow/examples/two_pc/coordinator.rs:152]: I believe the issue is that we generate a running count, and need to choose the highest number we ever issue. Yes? This won't be clear without a comment :-)

Comment (jhellerstein) [reply hydroflow/examples/two_pc/coordinator.rs:152]: Also note in the code that this is inefficient unless we do a lot of batching.

File: hydroflow/examples/two_pc/coordinator.rs [hunk 2]
```diff
@@ -41,24 +41,38 @@ pub(crate) async fn run_coordinator(opts: Opts, subordinates: Vec<String>) {
         hf.make_edge::<_, VecHandoff<String>, _>("subordinates_out_tee_2");
     let (subordinates_out_tee_3_push, subordinates_out_tee_3_pull) =
         hf.make_edge::<_, VecHandoff<String>, _>("subordinates_out_tee_3");
+    let (subordinates_out_tee_4_push, subordinates_out_tee_4_pull) =
+        hf.make_edge::<_, VecHandoff<String>, _>("subordinates_out_tee_4");
 
     // create a flow to populate the subordinate handoffs
-    hf.add_subgraph(
```
Comment (justinj) [inline hydroflow/examples/two_pc/coordinator.rs:46]: we need some syntax sugar for this, or to revisit `TeeingHandoff`, pretty gross you have to manually binarily tee like this

File: hydroflow/examples/two_pc/coordinator.rs [hunk 3]
```diff
@@ -118,38 +134,72 @@ pub(crate) async fn run_coordinator(opts: Opts, subordinates: Vec<String>) {
...
-                *vote_ctr.entry((msg.xid, msg.mtype)).or_insert(0) += 1;
-                // HACK: the constant 3 here is the length of the subordinates list!
-                // Should be computed on bootstrap tick and joined in here.
-                if vote_ctr.get(&(msg.xid, MsgType::Commit)).unwrap_or(&0)
-                    + vote_ctr.get(&(msg.xid, MsgType::Abort)).unwrap_or(&0)
-                    == 3
-                {
+            .map_scan(
+                HashMap::<u16, (u32, SubordResponse, bool)>::new(),
+                |txs, msg| {
```
Comment (jhellerstein) [inline hydroflow/examples/two_pc/coordinator.rs:183]: add comment and/or use more meaning  variable names so we understand the HashMap is holding ballot-counts for all transactions, partitioned by transactionId.

File: hydroflow/examples/two_pc/coordinator.rs [hunk 4]
```diff
@@ -172,24 +222,30 @@ pub(crate) async fn run_coordinator(opts: Opts, subordinates: Vec<String>) {
...
             })
-            .filter_map(move |msg| {
-                *ack_ctr.entry((msg.xid, msg.mtype)).or_insert(0) += 1;
-                // HACK
-                if ack_ctr.get(&(msg.xid, MsgType::AckP2)).unwrap_or(&0) + 0 == 3 {
-                    Some(msg)
-                } else {
-                    None
-                }
+            .map_scan(HashMap::<u16, (u32, SubordResponse)>::new(), |txs, msg| {
```
Comment (justinj) [inline hydroflow/examples/two_pc/coordinator.rs:234]: don't love the requirement to `clone` here, @MingweiSamuel wdyt about a `map_scan` that flushes its state once it's consumed everything, instead (and if you place it at a stratum boundary you know you have everything)?

File: hydroflow/examples/two_pc/coordinator.rs [hunk 5]
```diff
@@ -118,38 +134,72 @@ pub(crate) async fn run_coordinator(opts: Opts, subordinates: Vec<String>) {
...
-                // Should be computed on bootstrap tick and joined in here.
-                if vote_ctr.get(&(msg.xid, MsgType::Commit)).unwrap_or(&0)
-                    + vote_ctr.get(&(msg.xid, MsgType::Abort)).unwrap_or(&0)
-                    == 3
-                {
+            .map_scan(
+                HashMap::<u16, (u32, SubordResponse, bool)>::new(),
+                |txs, msg| {
+                    let commit = msg.mtype == MsgType::Commit;
+                    let v = txs.entry(msg.xid).or_insert((0, msg, commit));
```
Comment (jhellerstein) [inline hydroflow/examples/two_pc/coordinator.rs:185]: maybe add comment 
// make sure there's at least a bot-valued entry for this xid

File: hydroflow/examples/two_pc/coordinator.rs (hunk 4 above)
Comment (jhellerstein) [inline hydroflow/examples/two_pc/coordinator.rs:234]: similar to above: add comment and/or better variable name to clarify this is an ack counter per transaction.

File: hydroflow/examples/two_pc/members.json [hunk 6]
```diff
@@ -3,6 +3,7 @@
     "subordinates": [
         "localhost:12347",
         "localhost:12348",
-        "localhost:12349"
+        "localhost:12349",
+        "localhost:12350"
```
Comment (jhellerstein) [inline hydroflow/examples/two_pc/members.json:7]: revert this to 3 participants to make test startup simpler.

**Review by justinj — APPROVED** [review]
LGTM, @jhellerstein might want to take a look since he wrote the original code

**Review by jhellerstein — APPROVED** [review]

**Comment by justinj** [comment]
future work: we should just have a single rust program that starts up the coordinator and all the participants to simplify stuff.

File: hydroflow/examples/two_pc/coordinator.rs (hunk 4 above)
Comment (MingweiSamuel) [reply hydroflow/examples/two_pc/coordinator.rs:234]: Sounds like `fold()`. Has to be at stratum boundary of course

**Review by MingweiSamuel — APPROVED** [review]

File: hydroflow/examples/two_pc/coordinator.rs (hunk 5 above)
Comment (rithvikp) [reply hydroflow/examples/two_pc/coordinator.rs]: I'm not sure if I'm misunderstanding this comment: is it for the `or_insert`? Is a bot-valued entry the default entry that we add (with a `0` count)?

**Pull request merged**